### PR TITLE
SY-1016 Guarding write/iterate/stream endpoints with access control

### PR DESCRIPTION
**Pull request by LeonLiur**
# Feature Pull Request Template

## Key Information

- **Linear Issue**: [SY-1016](https://linear.app/synnaxlabs/issue/SY-1016/[server]-guarding-writeiteratestream-endpoints-with-access-control)

## Description

Introduced access checks on framer methods - iterate, write, stream, and delete.

## Basic Readiness Checklist

- [x] I have performed a self-review of my code.
- [x] I have added relevant tests to cover the changes to CI.
- [ ] I have updated user facing documentation accordingly.
- [ ] I have verified code coverage targets are met.

## Migrations

- [x] Console - I have ensured that previous versions of stored data structures are 
properly migrated to new formats.
- [x] Server - I have ensured that previous versions of stored data structures are 
properly migrated to new formats.

##  Additional Notes
- [ ] These changes deal with concurrency.
- [ ] These changes affect UI.

## Manual QA Additions

- [x] I have updated the [Release Candidate](/.github/PULL_REQUEST_TEMPLATE/rc.md) template
with necessary manual QA steps to test my change.

## Reviewer Checklist
- [ ] Sufficient test coverage of new additions.
- [ ] Verified all steps in the Readiness checklists.
- [ ] UI changes have been tested.
- [ ] Style and formatting is consistent.
- [ ] Reviewed any relevant changes to concurrent code for safety. 
- [ ] Sufficient comments and clarity of code.
- [ ] Any additional documentation necessary for feature is provided and has been reviewed for clarity.

### CODE REVIEW

File: synnax/pkg/api/api.go
Comment (pjdotson) [inline synnax/pkg/api/api.go:213]: @emilbon99 Will changing the name to `Framer` from `Telem` cause any problems with incompatibility?

File: synnax/pkg/api/api.go
Comment (pjdotson) [inline synnax/pkg/api/api.go:361]: You should change the handler to `Delete` instead of `FrameDelete`.

Comment (pjdotson) [reply synnax/pkg/api/api.go:361]: Also, there's a bunch of variables and functions in `framer.go` that have names that start with `Frame` or `Framer` - those names should be changed.

File: client/py/synnax/framer/streamer.py
Comment (pjdotson) [inline client/py/synnax/framer/streamer.py:59]: Replace with `_, exc`.

File: client/py/synnax/framer/streamer.py
Comment (pjdotson) [inline client/py/synnax/framer/streamer.py:125]: `exc` not `err`.

File: client/py/tests/test_access.py
Comment (pjdotson) [inline client/py/tests/test_access.py:15]: Have you checked that this pattern is matched now in schematic, range, line plot, etc.?

File: client/py/synnax/framer/streamer.py
Comment (pjdotson) [inline client/py/synnax/framer/streamer.py:35]: Is there a reason that it could be a string here?

**Review by pjdotson — CHANGES_REQUESTED** [review]

**Comment by Lham42** [comment]
To address your comments

1. I think it makes sense to decouple the framer and channel client 
2. No action/return is sufficient ack for success so I'd make sure you return error that can be checked rather than an ack
3. Once this pr has been approved, documentation can be updated to have a section on this for py and ts clients. The rest of the docs can be held off til UI is done.

**Comment by pjdotson** [comment]
On your comments:
1. @emilbon99 Do you have an idea on a better solution? This solves this problem but I think the problem is that channels are unique amongst all of our resources. I don't love the idea of making an ontology id for this but it seems preferable to adding separate actions for reading / updating data.
2. Shouldn't you send an error instead of an ack if a streamer failed to create? I feel like thats the more idiomatic solution in Go / matches the rest of the codebase better.
3. Yes lets add something on both Python / TypeScript client.

File: synnax/pkg/api/api.go
Comment (LeonLiur) [reply synnax/pkg/api/api.go:361]: This file is in `package api` and not `package framer`, so a lot of these variables do need to be prefixed with `frame` since otherwise they collide with other names in the `api` package, like channel delete.

File: client/py/synnax/framer/streamer.py
Comment (LeonLiur) [reply client/py/synnax/framer/streamer.py:125]: gotcha, also changed other occurances

File: client/py/tests/test_access.py
Comment (LeonLiur) [reply client/py/tests/test_access.py:15]: yep

File: client/py/synnax/framer/streamer.py
Comment (LeonLiur) [reply client/py/synnax/framer/streamer.py:35]: no, you're right, removed.

**Comment by LeonLiur** [comment]
> On your comments:
> 
> 1. @emilbon99 Do you have an idea on a better solution? This solves this problem but I think the problem is that channels are unique amongst all of our resources. I don't love the idea of making an ontology id for this but it seems preferable to adding separate actions for reading / updating data.
> 2. Shouldn't you send an error instead of an ack if a streamer failed to create? I feel like thats the more idiomatic solution in Go / matches the rest of the codebase better.
> 3. Yes lets add something on both Python / TypeScript client.

2. Yes, if there's an error, an empty frame + an error is sent; if there's no error, an empty frame + no error is sent. this way the `receive()` on the client side does not block.

**Review by Lham42 — CHANGES_REQUESTED** [review]

File: client/py/synnax/access/payload.py
Comment (emilbon99) [inline client/py/synnax/access/payload.py:1]: Do we still need the ALLOW_ALL here?

File: client/py/synnax/channel/client.py
Comment (emilbon99) [inline client/py/synnax/channel/client.py:1]: The formatting on these elipses is very strange

File: client/py/synnax/framer/writer.py
Comment (emilbon99) [inline client/py/synnax/framer/writer.py:1]: Weird formatting on elipses

File: client/ts/src/access/policy/payload.ts
Comment (emilbon99) [inline client/ts/src/access/policy/payload.ts:1]: Do we still need the ALLOW_ALL type?

**Review by emilbon99 — APPROVED** [review]
LGTM once these comments are addressed.

**Pull request merged**
